### PR TITLE
feat(workspace/desktop): add new resource support

### PR DESCRIPTION
**Pull request by Lance52259**
**What this PR does / why we need it**:
Add a new Workspace resource support.

**Which issue this PR fixes**:
*(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*
fixes #xxx

**Special notes for your reviewer**:

**Release note**:

```release-note
1. add a new workspace resource support.
```

## PR Checklist

* [x] Tests added/passed.
* [x] Documentation updated.
* [x] Schema updated.

## Acceptance Steps Performed

```
make testacc TEST='./huaweicloud/services/acceptance/workspace' TESTARGS='-run=TestAccDesktop_basic'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud/services/acceptance/workspace -v -run=TestAccDesktop_basic -timeout 360m -parallel 4
=== RUN   TestAccDesktop_basic
=== PAUSE TestAccDesktop_basic
=== CONT  TestAccDesktop_basic
--- PASS: TestAccDesktop_basic (497.84s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/workspace 497.891s
```

### CODE REVIEW

File: docs/resources/workspace_desktop.md [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+* `root_volume` - (Required, List) Specifies the configuration of system volume.
+  The [object](#desktop_volume) structure is documented below.
+
+* `data_volume` - (Optional, List) Specifies the configuration of data volumes.
+  The [object](#desktop_volume) structure is documented below.
+
+* `availability_zone` - (Optional, String, ForceNew) Specifies the availability zone where the desktop is located.
+  Changing this will create a new resource.
+
+* `security_groups` - (Optional, String, ForceNew) Specifies the ID list of security groups.
```
Comment (Jason-Zhang9309) [inline docs/resources/workspace_desktop.md]: this is not a string

File: docs/resources/workspace_desktop.md [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+  In addition to the custom security group, it must also contain a security group called **WorkspaceUserSecurityGroup**.
+  Changing this will create a new resource.
+
+* `user_group` - (Required), String, ForceNew) Specifies the user group to which the desktop belongs.
```
Comment (Jason-Zhang9309) [inline docs/resources/workspace_desktop.md]: extra `)`, and how about move this to line 89

File: huaweicloud/services/workspace/resource_huaweicloud_workspace_desktop.go [hunk 3]
```diff
@@ -0,0 +1,622 @@
...
+				Optional: true,
+				ForceNew: true,
+			},
+			"tags": {
+				Type:     schema.TypeMap,
+				Optional: true,
+				Computed: true,
+				ForceNew: true,
+				Elem:     &schema.Schema{Type: schema.TypeString},
+			},
```
Comment (Jason-Zhang9309) [inline huaweicloud/services/workspace/resource_huaweicloud_workspace_desktop.go]: you can use common.TagsForceNewSchema

File: docs/resources/workspace_desktop.md [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+
+* `email_notification` - (Optional, Bool, ForceNew) Specifies whether to send emails to user mailbox during important
+  operations. Changing this will create a new resource.
+
+* `tags` - (Optional, Map, ForceNew) Specifies the key/value pairs of the desktop.
+  Changing this will create a new resource.
+
+* `delete_user` - (Optional, Bool) Specifies whether to delete user associated with this desktop after deleting it.
+  The user can only be successfully deleted if the user has no other desktops.
+  Changing this will create a new resource.
```
Comment (Jason-Zhang9309) [inline docs/resources/workspace_desktop.md]: it's not forcenew in the schema

File: docs/resources/workspace_desktop.md [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+  + For root volume, the valid value is range from `80` to `1,020`.
+  + For data volume, the valid value is range from `10` to `8,200`.
+
+<a name="desktop_nic"></a>
+The `nic` block supports:
+
+* `network_id` - (Required, String, ForceNew) Specifies the network ID.
+  The valid values are as follows:
+  + **SAS**: High I/O disk type.
+  + **SSD**: Ultra-high I/O disk type.
```
Comment (Jason-Zhang9309) [inline docs/resources/workspace_desktop.md]: the values are not correct, and it's not forcenew in the schema

File: huaweicloud/services/acceptance/workspace/resource_huaweicloud_workspace_desktop_test.go [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+					resource.TestCheckResourceAttr(resourceName, "data_volume.0.type", "SAS"),
+					resource.TestCheckResourceAttr(resourceName, "data_volume.0.size", "50"),
+					resource.TestCheckResourceAttr(resourceName, "data_volume.1.type", "SAS"),
+					resource.TestCheckResourceAttr(resourceName, "data_volume.1.size", "90"),
+					resource.TestCheckResourceAttr(resourceName, "data_volume.2.type", "SAS"),
+					resource.TestCheckResourceAttr(resourceName, "data_volume.2.size", "20"),
+					resource.TestCheckResourceAttr(resourceName, "data_volume.3.type", "SAS"),
+					resource.TestCheckResourceAttr(resourceName, "data_volume.3.size", "40"),
+				),
+			},
```
Comment (Jason-Zhang9309) [inline huaweicloud/services/acceptance/workspace/resource_huaweicloud_workspace_desktop_test.go:77]: add an import test

File: docs/resources/workspace_desktop.md (hunk 1 above)
Comment (Lance52259) [reply docs/resources/workspace_desktop.md]: fixed.

File: docs/resources/workspace_desktop.md (hunk 2 above)
Comment (Lance52259) [reply docs/resources/workspace_desktop.md]: fixed.

File: huaweicloud/services/workspace/resource_huaweicloud_workspace_desktop.go (hunk 3 above)
Comment (Lance52259) [reply huaweicloud/services/workspace/resource_huaweicloud_workspace_desktop.go]: done.

File: docs/resources/workspace_desktop.md (hunk 4 above)
Comment (Lance52259) [reply docs/resources/workspace_desktop.md]: the wrong description has been deleted.

File: docs/resources/workspace_desktop.md (hunk 5 above)
Comment (Lance52259) [reply docs/resources/workspace_desktop.md]: fixed.

File: huaweicloud/services/acceptance/workspace/resource_huaweicloud_workspace_desktop_test.go (hunk 6 above)
Comment (Lance52259) [reply huaweicloud/services/acceptance/workspace/resource_huaweicloud_workspace_desktop_test.go:77]: finished

**Comment by Lance52259** [comment]
```
make testacc TEST='./huaweicloud/services/acceptance/workspace' TESTARGS='-run=TestAccDesktop_basic'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud/services/acceptance/workspace -v -run=TestAccDesktop_basic -timeout 360m -parallel 4
=== RUN   TestAccDesktop_basic
=== PAUSE TestAccDesktop_basic
=== CONT  TestAccDesktop_basic
--- PASS: TestAccDesktop_basic (753.31s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/workspace        753.366s
```

File: huaweicloud/config/config.go [hunk 7]
```diff
@@ -637,6 +637,10 @@ func (c *Config) DnsWithRegionClient(region string) (*golangsdk.ServiceClient, e
 	return c.NewServiceClient("dns_region", region)
 }
 
+func (c *Config) WorkspaceV2Client(region string) (*golangsdk.ServiceClient, error) {
```
Comment (ShiChangkuo) [inline huaweicloud/config/config.go]: suggest to add an acc-test in huaweicloud/endpoint_test.go

File: huaweicloud/services/workspace/resource_huaweicloud_workspace_desktop.go [hunk 8]
```diff
@@ -0,0 +1,616 @@
...
+			"security_groups": {
+				Type:     schema.TypeSet,
+				Optional: true,
+				Computed: true,
+				ForceNew: true,
+				Elem:     &schema.Schema{Type: schema.TypeString},
+			},
+			"user_group": {
+				Type:     schema.TypeString,
+				Required: true,
```
Comment (ShiChangkuo) [inline huaweicloud/services/workspace/resource_huaweicloud_workspace_desktop.go]: please define this parameter after user_email as it is required

File: huaweicloud/services/workspace/resource_huaweicloud_workspace_desktop.go [hunk 9]
```diff
@@ -0,0 +1,616 @@
...
+				},
+			},
+			"name": {
+				Type:          schema.TypeString,
+				Optional:      true,
+				Computed:      true,
+				ForceNew:      true,
+				ConflictsWith: []string{"name_prefix"},
+			},
+			"name_prefix": {
```
Comment (ShiChangkuo) [inline huaweicloud/services/workspace/resource_huaweicloud_workspace_desktop.go]: it seems that name_prefix is not suitable for terraform, keep `name` will be ok

File: huaweicloud/services/workspace/resource_huaweicloud_workspace_desktop.go [hunk 10]
```diff
@@ -0,0 +1,616 @@
...
+			"type":       volume.Type,
+			"size":       volume.Size,
+			"id":         volume.VolumeId,
+			"name":       volume.Name,
+			"device":     volume.Device,
+			"created_at": volume.CreatedAt,
+		}
+	}
+
+	// Since the volumes in the response body are unordered, they are sorted by name.
```
Comment (ShiChangkuo) [inline huaweicloud/services/workspace/resource_huaweicloud_workspace_desktop.go]: s/name/device

File: huaweicloud/services/acceptance/workspace/resource_huaweicloud_workspace_desktop_test.go [hunk 11]
```diff
@@ -0,0 +1,219 @@
...
+			{
+				ResourceName:      resourceName,
+				ImportState:       true,
+				ImportStateVerify: true,
+				ImportStateVerifyIgnore: []string{
+					"delete_user",
+					"image_type",
+					"nic.#",
+					"nic.0.%",
+					"nic.0.network_id",
```
Comment (ShiChangkuo) [inline huaweicloud/services/acceptance/workspace/resource_huaweicloud_workspace_desktop_test.go]: can just ignore "nic" ?

File: huaweicloud/config/config.go (hunk 7 above)
Comment (Lance52259) [reply huaweicloud/config/config.go]: done.

File: huaweicloud/services/workspace/resource_huaweicloud_workspace_desktop.go (hunk 8 above)
Comment (Lance52259) [reply huaweicloud/services/workspace/resource_huaweicloud_workspace_desktop.go]: done.

File: huaweicloud/services/workspace/resource_huaweicloud_workspace_desktop.go (hunk 9 above)
Comment (Lance52259) [reply huaweicloud/services/workspace/resource_huaweicloud_workspace_desktop.go]: done.

File: huaweicloud/services/workspace/resource_huaweicloud_workspace_desktop.go (hunk 10 above)
Comment (Lance52259) [reply huaweicloud/services/workspace/resource_huaweicloud_workspace_desktop.go]: done.

File: huaweicloud/services/acceptance/workspace/resource_huaweicloud_workspace_desktop_test.go (hunk 11 above)
Comment (Lance52259) [reply huaweicloud/services/acceptance/workspace/resource_huaweicloud_workspace_desktop_test.go]: sure, it works.

**Comment by Lance52259** [comment]
```
make testacc TEST='./huaweicloud/services/acceptance/workspace' TESTARGS='-run=TestAccDesktop_basic'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud/services/acceptance/workspace -v -run=TestAccDesktop_basic -timeout 360m -parallel 4
=== RUN   TestAccDesktop_basic
=== PAUSE TestAccDesktop_basic
=== CONT  TestAccDesktop_basic
--- PASS: TestAccDesktop_basic (765.81s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/workspace 765.879s
```

**Pull request merged**